packages/trie/src/trie.ts

```diff
@@ -770,22 +814,19 @@ export class Trie {
    * Saves the nodes from a proof into the trie.
    * @param proof
    */
-  async fromProof(proof: Proof): Promise<void> {
+
+  async updateFromProof(proof: Proof): Promise<void> {
+    if (!equalsBytes(this.root(), this.hash(proof.shift()!))) {
```

Comment:
I am not sure if I like this check, have to think about the practical downsides of this :thinking:

Reply:
If you try to update a sparse trie with an invalid proof, this will throw.
shift() because the root node is already in the trie, so we'll never need to add it.

Reply:
:thinking:  I'd have to hear the upside to allowing a trie to update itself from a proof to a different trie!